client/proposer-metrics/src/lib.rs

```diff
@@ -0,0 +1,69 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Prometheus basic proposer metrics.
+
+use prometheus_endpoint::{register, PrometheusError, Registry, Histogram, HistogramOpts, Gauge, U64};
+
+/// Optional shareable link to basic authorship metrics.
+#[derive(Clone, Default)]
+pub struct MetricsLink(Option<Metrics>);
```

Comment:
Why this additional wrapper around `Metrics`?

Reply:
So that we can avoid `if let Some(metrics) = self.metrics { ... }` everywhere and just use `metrics.report(|metrics| ...)`

Reply:
Sounds good to me. I guess you could as well do `metrics.as_ref().map(|metrics| metrics.some_counter.inc())` but that would violate the assumption of a `map` function being pure.